File: tests/test_hmac.py
"""Tests for HMAC verification and internal signature replay protection."""

import hashlib
import hmac as stdlib_hmac
import time

import pytest

from monkey.hmac import (
    BadSignature,
    hmac_sign,
    verify_github_signature,
    verify_internal_signature,
)

SECRET = "test-secret"
BODY = b'{"action":"opened","issue":{"number":123}}'


def test_valid_github_signature_passes():
    sig = hmac_sign(SECRET, BODY)
    verify_github_signature(SECRET, BODY, sig)  # no raise


def test_wrong_secret_raises():
    sig = hmac_sign("other-secret", BODY)
    with pytest.raises(BadSignature):
        verify_github_signature(SECRET, BODY, sig)


def test_tampered_body_raises():
    sig = hmac_sign(SECRET, BODY)
    with pytest.raises(BadSignature):
        verify_github_signature(SECRET, BODY + b"x", sig)


def test_missing_header_raises():
    with pytest.raises(BadSignature):
        verify_github_signature(SECRET, BODY, None)


def test_unsupported_scheme_raises():
    with pytest.raises(BadSignature):
        verify_github_signature(SECRET, BODY, "sha1=abc")


def test_internal_signature_within_skew_passes():
    ts = int(time.time())
    sig = hmac_sign(SECRET, BODY, ts)
    verify_internal_signature(SECRET, BODY, sig, timestamp_header=str(ts))


def test_internal_signature_replay_rejected():
    ts = int(time.time()) - 100  # outside ±30s skew
    sig = hmac_sign(SECRET, BODY, ts)
    with pytest.raises(BadSignature):
        verify_internal_signature(SECRET, BODY, sig, timestamp_header=str(ts))


def test_internal_signature_replay_with_fresh_timestamp_rejected():
    """Regression: a captured (body, signature) pair must not verify when the
    attacker refreshes x-monkey-ts inside the ±skew window."""
    captured_ts = int(time.time())
    captured_sig = hmac_sign(SECRET, BODY, captured_ts)
    fresh_ts = captured_ts + 10  # still inside ±30s skew
    with pytest.raises(BadSignature):
        verify_internal_signature(
            SECRET, BODY, captured_sig, timestamp_header=str(fresh_ts)
        )


def test_internal_signature_without_timestamp_binding_rejected():
    """Regression: a signature computed over the raw body alone (timestamp not
    bound in) must not pass verification."""
    ts = int(time.time())
    body_only_sig = "sha256=" + stdlib_hmac.new(
        SECRET.encode(), BODY, hashlib.sha256
    ).hexdigest()
    with pytest.raises(BadSignature):
        verify_internal_signature(SECRET, BODY, body_only_sig, timestamp_header=str(ts))

File: tests/test_hmac.rs
use monkey::hmac_auth::{
    HmacError, hmac_sign, hmac_sign_with_timestamp, verify_github_signature,
    verify_internal_signature,
};
use std::time::{SystemTime, UNIX_EPOCH};

const SECRET: &str = "test-secret";
const BODY: &[u8] = b"{\"action\":\"opened\",\"issue\":{\"number\":123}}";

#[test]
fn test_valid_github_signature_passes() {
    let sig = hmac_sign(SECRET, BODY);
    assert!(verify_github_signature(SECRET, BODY, Some(&sig)).is_ok());
}

#[test]
fn test_wrong_secret_raises() {
    let sig = hmac_sign("other-secret", BODY);
    assert_eq!(
        verify_github_signature(SECRET, BODY, Some(&sig)),
        Err(HmacError::SignatureMismatch)
    );
}

#[test]
fn test_tampered_body_raises() {
    let sig = hmac_sign(SECRET, BODY);
    let mut tampered = BODY.to_vec();
    tampered.push(b'x');
    assert_eq!(
        verify_github_signature(SECRET, &tampered, Some(&sig)),
        Err(HmacError::SignatureMismatch)
    );
}

#[test]
fn test_missing_header_raises() {
    assert_eq!(
        verify_github_signature(SECRET, BODY, None),
        Err(HmacError::MissingSignature)
    );
}

#[test]
fn test_unsupported_scheme_raises() {
    assert_eq!(
        verify_github_signature(SECRET, BODY, Some("sha1=abc")),
        Err(HmacError::UnsupportedScheme)
    );
}

#[test]
fn test_internal_signature_within_skew_passes() {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;
    let sig = hmac_sign_with_timestamp(SECRET, BODY, ts);
    assert!(verify_internal_signature(SECRET, BODY, Some(&sig), Some(&ts.to_string()), 30).is_ok());
}

#[test]
fn test_internal_signature_replay_rejected() {
    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        - 100; // outside +-30s skew
    let sig = hmac_sign_with_timestamp(SECRET, BODY, ts);
    assert_eq!(
        verify_internal_signature(SECRET, BODY, Some(&sig), Some(&ts.to_string()), 30),
        Err(HmacError::TimestampSkew)
    );
}

#[test]
fn test_internal_signature_rejects_fresh_timestamp_replay() {
    let old_timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        - 3600;
    let fresh_timestamp = old_timestamp + 3600;
    let captured_signature = hmac_sign_with_timestamp(SECRET, BODY, old_timestamp);

    assert_eq!(
        verify_internal_signature(
            SECRET,
            BODY,
            Some(&captured_signature),
            Some(&fresh_timestamp.to_string()),
            30,
        ),
        Err(HmacError::SignatureMismatch)
    );
}

#[test]
fn test_internal_signature_rejects_unbound_signature() {
    let timestamp = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64;
    let unbound_signature = hmac_sign(SECRET, BODY);

    assert_eq!(
        verify_internal_signature(
            SECRET,
            BODY,
            Some(&unbound_signature),
            Some(&timestamp.to_string()),
            30,
        ),
        Err(HmacError::SignatureMismatch)
    );
}
